launchpad/launchpad.gno

```diff
@@ -0,0 +1,399 @@
+		"startTime", formatUint(startTime),
+		"startHeight", formatUint(startHeight),
+		"projectId", projectId,
+		"tier30Amount", formatUint(calcResult.Tier30Amount),
+		"tier30EndHeight", formatUint(tier30.ended.height),
+		"tier90Amount", formatUint(calcResult.Tier90Amount),
+		"tier90EndHeight", formatUint(tier90.ended.height),
+		"tier180Amount", formatUint(calcResult.Tier180Amount),
+		"tier180EndHeight", formatUint(tier180.ended.height),
+	)
+
+	return projectId
+}
+
+// middle result of project params calculation
+type ProjectCalculationResult struct {
+	Tier30Amount  uint64
+	Tier90Amount  uint64
+	Tier180Amount uint64
+	StartHeight   uint64
+}
+
+func calculateProjectParams(input ProjectInput, now uint64) (*ProjectCalculationResult, error) {
+	tier30Amount := input.DepositAmount * input.Tier30Ratio / 100
+	tier90Amount := input.DepositAmount * input.Tier90Ratio / 100
+	tier180Amount := input.DepositAmount * input.Tier180Ratio / 100
+
+	if input.StartTime < now {
+		return nil, ufmt.Errorf("startTime(%d) must be greater than now(%d)", input.StartTime, now)
+	}
+
+	// calculate when the block starts
+	timeUntilStart := input.StartTime - now
+
+	avgBlockTimeMs := uint64(gns.GetAvgBlockTimeInMs())
+	if avgBlockTimeMs == 0 {
+		return nil, ufmt.Errorf("average block time not available")
+	}
+
+	blockDuration := timeUntilStart * 1000 / avgBlockTimeMs
+	height := uint64(std.GetHeight())
+	startHeight := height + blockDuration
+
+	return &ProjectCalculationResult{
+		Tier30Amount:  tier30Amount,
+		Tier90Amount:  tier90Amount,
+		Tier180Amount: tier180Amount,
+		StartHeight:   startHeight,
+	}, nil
+}
+
+func covertToMillisecond(timestamp uint64) uint64 {
+	return timestamp * 1000 / uint64(gns.GetAvgBlockTimeInMs())
+}
+
+// validateTransferLeft validates the transfer of remaining tokens
+func validateTransferLeft(project Project, recipient std.Address, height uint64) error {
+	if !recipient.IsValid() {
+		return errors.New(ufmt.Sprintf("invalid recipient address(%s)", recipient.String()))
+	}
+
+	if height < project.ended.height {
+		return errors.New(ufmt.Sprintf("project not ended yet(current:%d, endHeight: %d)",
+			height, project.ended.height))
+	}
+
+	if project.refund.height != 0 {
+		return errors.New(ufmt.Sprintf("project already refunded(height:%d)",
+			project.refund.height))
+	}
+
+	return nil
+}
+
+// calculateLeftReward calculates the remaining reward amount for each tier
+func calculateLeftReward(project Project) uint64 {
+	tier30 := project.tiers[30]
+	tier90 := project.tiers[90]
+	tier180 := project.tiers[180]
+
+	left30 := tier30.tierAmount - tier30.calculatedAmount
+	left90 := tier90.tierAmount - tier90.calculatedAmount
+	left180 := tier180.tierAmount - tier180.calculatedAmount
+	return left30 + left90 + left180
+}
+
+func TransferLeftFromProjectByAdmin(projectId string, recipient std.Address) uint64 {
+	if err := common.SatisfyCond(isUserCall()); err != nil {
+		panic(addDetailToError(errAlreadyRefunded, err.Error()))
```

Comment:
error message seems to be wrong.
checker logic(isUserCall()) has nothing to do with `already refund`